### ISSUE TITLE
Enabling access to intended recipients 

### ISSUE THREAD

**Issue by jelordreygulle** [virtru/encrypt-decrypt-demo#43]
Securing a file such that only intended recipients can access its data. How do we set the emails of people we wanted to allow or send files to ?

**Comment by jrschumacher** [comment]
Hi @jelordreygulle thanks for asking. Here's a bit of code that will do what you want.

```js
  // Create your client
  const client = new Virtru.Client();

  // Create a new policy instructions. The policy builder will be a object instructing the policy needs
  const policy = new Virtru.PolicyBuilder()
    // Let's add you to the policy. This will allow you to decrypt the policy.
    .addUsersWithAccess(["jelordreygulle@example.com"])
    // Let's build the policy object and and store it to `policy`
    .build();

  // Create encrypted instructions. The EncryptParamsBuilder creates an object instructing what encryption needs are
  const encryptParams = new Virtru.EncryptParamsBuilder()
    // Which file are we encrypting?
    .withFileSource('./file-to-encrypt.txt')
    // What will the file be called when decrypted?
    .withDisplayFilename('secrets.txt')
    // Attach the policy above
    .withPolicy(policy)
    // Build the encrypted params
    .build();

  // Encrypt the file to a stream
  const protectedStream = await client.encrypt(encryptParams);

  // Name the file
  const protectedFile = 'secrets.tdf.html';
  await protectedStream
    .toFile(`./${protectedFile}`)
    .then(() => console.log(`Encrypted file ${protectedFile}`));
```

**Comment by jrschumacher** [comment]
@jelordreygulle if the data is owned by you (i.e. you created the data and sharing it with customers) then you want to use the Node API. If the data is owned by your customers (i.e. your customers have created the data and you are sharing it with someone) then you want to use the browser code.

Could you describe the product that you're developing? We might be able to help you figure out where Virtru SDK can be added.

**Comment by jelordreygulle** [comment]
@jrschumacher  Thank you for replying Sir , I do really appreciate it . My goal is that

A user goes to the “site” we will have like my.testsite..com or something and it will ask them to authenticate if needed.  Ideally the sending user doesn’t have to authenticate or if so it is easy.  Then it will allow them drag the files they want to send and enter the emails of the people they wish to send the files to.  The send to will only allow certain emails with extensions ,  all others will fail.  

Then the receiving party will receive an email with a means of downloading and decrypting the files. I prefer to use end to end encryption Sir(NOT use their servers to encrypt and de-crypt the files).

The files need to store on AWS.  All API’s are API gateway.  All logic is lambda.  All storage is S3. 

Site storage is s3 with cloud front distribution.  We use serverless.com framework to manage it all. 

Hoping for you reply Sir @jrschumacher , much appreciated. Thank you.

**Comment by jrschumacher** [comment]
@jelordreygulle okay so you want to do this in the browser. This way you never touch the customers files and only the encrypted data. This will ensure your customers own the data and you can respect their privacy.

**Comment by jelordreygulle** [comment]
> @jelordreygulle okay so you want to do this in the browser. This way you never touch the customers files and only the encrypted data. This will ensure your customers own the data and you can respect their privacy.

I see , so it would be best to use the client side sdk Sir ? , what can you suggest when it comes to the node sdk Sir regarding on my stated/described above. Thanks

**Comment by jelordreygulle** [comment]
Also about storing the files on aws s3 , do you have any advice or suggestions Sir ? Thanks

**Comment by jrschumacher** [comment]
Yes. So a proposed usecase:

1. User loads example.com
2. User authenticates with example.com
3. User proceeds to share file section
4. User adds the files they want to share
5. User specifies user they want to share with
6. Application prompts for Virtru authentication
7. User authenticates with Virtru
8. User presses encrypt and send button
9. Virtru SDK creates a policy which shares with User provided email address and encrypts the data
10. Application requests [presigned AWS S3 url ](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getSignedUrl-property)
11. Application sends XMLHttpRequest to AWS presigned URL with encrypted document
12. Application does final chores
13. User rejoices 🎉

**Comment by jelordreygulle** [comment]
Thank you sir . Just one more clarification Sir , theres no need for me to use the node sdk right ? since the data is owned by the customer.

**Comment by jrschumacher** [comment]
@jelordreygulle you may want to look at these samples: https://github.com/virtru/virtru-sdk-js-samples

**Comment by jelordreygulle** [comment]
how does receiving user download and decrypt their files
Do they get a link that shows the file names and has a decrypt button ?

**Comment by jrschumacher** [comment]
Yep that works. You only need to implement the JS SDK and the AuthWidget.

**Comment by jelordreygulle** [comment]
Btw could no longer locate the virtru-sdk-js-samples/virtru-sdk-js-webform from the Virtru SDK Secure Form and File Submission example you have given.

**Comment by jrschumacher** [comment]
It got renamed. https://github.com/virtru/virtru-sdk-samples-js/tree/master/upload-form
